### DISCUSSION TITLE
Course projects

### DISCUSSION THREAD

**Discussion by patham9** [AGI-course/agi-course.github.io#1] [Announcements]
Hi everyone!

As you know next week on October 20 there will be no lecture.
We will take this as an opportunity to get the discussions about course projects started.
Please feel free to create a forum post here in the Ideas tab (or https://github.com/orgs/AGI-course/teams/project-discussions/discussions if you prefer it to be readable only by added Github users), containing:

Title: Name of project
Type: Software / Writing
Overview: What it is about, containing what could be done or written in the project

This will allow to comment on each others project ideas and for us to make suggestions so that it will be feasible to carry out within the time frame of this course.

For writing projects:
- Due to the multi-disciplinary nature of this field, you are of course very welcome to take inspirations from other related fields.

For software projects:
- If architecture-centered: Capabilities you want your AI system/model to have.
- If application-centered: You might want to think about cases where there are benefits of having general-purpose AI over systems only programmed and trained for a single purpose.

Best regards,
Patrick

**Comment by BirgerMoell** [comment]
Title: Artificial Humans
Type: Software
Overview: The project explores creating artificial humans using open domain chatbot systems and comparing them to the humans themselves in conversation. The project will evaluate participants interacting with both the artificial system and the human in terms of likability, flow of conversation and overall engagement in the conversation. The users will answer questions such as:

- How engaging was this conversation?
- How intelligent would you rate the person you are speaking with?
- How human-like would you rate this conversation?
- What was your favorite thing about this conversation?
- What was the thing you liked the least with this conversation?`

**Reply by patham9** [reply]
Do you plan to include some AGI-related capabilities in this project, such as allowing the chatbot to perform inductive reasoning / to let it maintain a long-term memory of the conversation it can then be asked about?
If so, how do you plan to extend the available technology to allow for that?
Will it be special AIML patterns which will trigger knowledge addition, e.g. "my * is green", which will, from "my hat is green" extract **hat --> green_thing** to allow for reasoning and query answering?

**Reply by BirgerMoell** [reply]
@patham9 I will use a deep learning system for chatbots that includes long term memory. The system has the ability to answer natural language question and reason about them.

**Reply by patham9** [reply]
Thanks for clarifying! It will be interesting how the solution deals with memory augmentation, as Transformer models only provide a fixed-sized contextual window.

**Comment by dwanev** [comment]
Title: Speech Recognition with Non-Axiomatic-Reasoning
Type: Software

Overview: Explore perception and NARs, or specifically speech recognition. Speech recognition will require the temporal portions of NARs as well as the other capabilities.

RQ1: Can non axiomatic reasoning, which can cope with conflicting information, be leveraged to perform speech recognition? 
RQ2: What level of performance can be obtained 
RQ3: Is there a computational advantage compared with other techniques? 
RQ4: Can the NARs be leveraged to provide explanations?

**Reply by patham9** [reply]
Very interesting! Here is some experimental perceptual reasoning capability which could potentially be valuable for your work:
https://github.com/opennars/OpenNARS-for-Applications/blob/master/misc/Python/Nalifier_ex1.nal
https://github.com/opennars/OpenNARS-for-Applications/blob/master/misc/Python/Nalifier_ex2.nal
https://github.com/opennars/OpenNARS-for-Applications/blob/master/misc/Python/Nalifier_ex3.nal
https://github.com/opennars/OpenNARS-for-Applications/blob/master/misc/Python/Nalifier_ex4.nal
The ideas are summarized in: https://docs.google.com/presentation/d/1x37XOr2B-cUlQS2_GQ6z7_uiYtAyTK2aN_Nx2G1hGog/edit?usp=sharing
Since perception requires NAL 1-3 inference to be carried out on a large scale, these examples do not work well with the generic control mechanisms of OpenNARS and ONA. Instead, they require a custom control mechanism to work well, you can run them from the /misc/Python/ directory of ONA using
`python3 Nalifier.py < Nalifier_ex1.nal | ./../../NAR shell | python3 ./../../colorize.py`
Nalifier serves the purpose of large-scale pattern matching and category formation for sensory channels.

**Reply by dwanev** [reply]
Great, will take a look. Many thanks!

**Comment by ArniDagur** [comment]
**Title**: Inference over continuous variables
**Type**: Software

### Overview

Architecture-centered. Given an environment whose variables are:

* Few in number, e.g. `1` to `10`.
* Continuous, with a range `[-∞, ∞]` or `[-1, 1]`.
* Mostly independent / uncorrelated.

Real life will never give you a low dimensional space with independent basis vectors, outright. However, the idea is that Deep Learning could help craft such a low dimensional space. See the literature on Disentangled Representations (for example by Higgins et. al.). If given such representation, the idea is that AGI-like reasoning might be applicable and desirable. 

Namely, system should able to perform simple temporal reasoning or inference:

* When I'm at `X` position in state space, `Y` tends to happen next.
* I'm at `Y` position, so maybe `X` happened before.

The consistency of forwards and backwards reasoning could act as a regularization or learning signal. Do this while:

* Learning in real time (no training / inference split)
* Mananging own time and memory. AIKR.

Just predicting the future is somewhat boring. I think a discussion with Patrick might shed light on how to make this assignment interesting from the perspective of AGI while respecting the time limit we have.

**Reply by ArniDagur** [reply]
I wasn't able to achieve the results I wanted given the idea put forth above. Thus, I started a writing project instead, which is around half way done. In the essay, I will talk about the importance and benefits of using a continuous representation in an AGI system, while also talking about the difficulties in working with continuous values.

**Comment by Xuelong-Fan** [comment]
Title: Explore inference reasoning in epidemiology
Type: Software – application-centered

Background: 
Association is a common relation of factors used in epidemiology to identify potential risk factors for certain adverse health outcomes. Traditional statistical methods are based on correlation or chi-square tests, and p-values. NARS is good at inference reasoning, but the results from NARS are reflected as frequency and confidence. However, it is yet unknown if NARS can be used to assess associations in epidemiological datasets. It is not clear how one can compare the results of the statistical methods with NARS, either. 

This project aims to:
1. construct a script-based interface that can assess associations of potential risk factors and a health outcome in a sample dataset about hyponatremia. 
2. attempt to find meaningful and clinically relevant interpretations of results from the NARS reasoning. 
3. compare results between NARS and traditional statistical methods.

Expected outcomes:
1. a script-based interface.
2. a scientific report.

**Comment by sofiathunberg** [comment]
Title: AGI and Robot Cognition
Type: Writing
Overview: Compare NARS with Brian Cantwell Smith book "The Promise of Artificial Intelligence" and other AI philosophers or roboethicist.

**Comment by AliBeikmohammadi** [comment]
Title: NARS vs Reinforcement learning 
Type: Software

Overview: One of the realistic scenarios is taking a sequence of optimal actions to do a task. Reinforcement learning is the most well-known approach to deal with this kind of task in the machine learning community. Finding a suitable alternative could always be an interesting and out-of-the-box matter. Therefore, in this project, we are looking to investigate the capability of NARS and answer the question of whether NARS has the potential to be a substitute for RL or not. Particularly, we are planning to make a comparison between RL and NARS on some environments developed by an Open AI gym. In this regard, the OpenNARS framework will be used.

**Comment by erl-j** [comment]
Working Title:  Automating Artistic Creativity
Type: Essay

Overview: This essay attempts to frame artistic creativity as a slightly unusual optimization problem. More details will follow.

**Comment by lcrosvila** [comment]
Title: Music understanding from an AGI perspective
Type: Writing
Overview: This paper attempts to frame the problem of music listening as a multi-level abstraction process where sensory information is represented as concepts, which can be updated and relate to each other. We discuss a variety of weaknesses of state-of-the-art music listening machines and consider how some of those can be managed through AGI principles. Finally, we outline how such a music listening machine could be.

**Comment by shubham2279** [comment]
Title: Intelligent decision-making in Wireless Edge Networks using NARS, ML, and traditional approaches
Type: Software

Overview: In Wireless edge networks, the edge nodes need to make several intelligent decisions like path planning, resource allocation, data offloading, etc. The goal of this project is to utilize Open-NARS to empower an edge node with some intelligent decision-making capabilities. A comparison of the performance of NARS-based decision-making will be done with some existing Machine learning and traditional algorithms to compare their effectiveness in solving problems related to Wireless edge networks.

**Comment by annchong** [comment]
Title: Automated Feature Engineering using NARS
Type: Software
Overview: Want to investigate the potential to leverage NARS for automating the feature engineering process.  

More details to follow.

**Comment by meltemgureller** [comment]
Title: IS IT STILL POSSIBLE FOR HUMANITY AND ARTIFICIAL INTELLIGENCE: ETHICAL SIDE
Type: Writing
Overview: The study is an example of a review article in which the concept of ethics in artificial intelligence systems is discussed. Although there are more than three hundred ethical documents in force for reliable artificial intelligence studies algorithmic bias, etc. the threats are still ongoing. A general ethical manifesto effort for artificial intelligence systems in provable security is the main topic of discussion in this article.
Meltem Güreller

**Comment by dwanev** [comment]
Hi Meltem,

I hope you are enjoying the course as much as I am. I just became aware of
a government initiative to influence this area. I do not know much about
what they are doing, but have linked a job description and link that should
lead you to other relevant information if it is of use.

kind regards

Dwane

Job summary
The Centre for Data Ethics and Innovation (CDEI) leads the UK government’s
work to enable trustworthy innovation using data and AI. We partner with
organisations across the public sector, industry and academia to develop
reliable approaches to data and AI governance, and tackle barriers to
growing innovation. We are looking for policy advisors to join CDEI’s
projects team, supporting the delivery of its programmes. One of the roles
will be specific to the CDEI’s AI Assurance Programme, which supports the
Government’s ambition to build the most trusted and pro-innovation AI
governance system in the world. You will be able to indicate any preference
you may have at interview.


https://www.civilservicejobs.service.gov.uk/csr/index.cgi?SID=c2VhcmNocGFnZT0xJnBhZ2VhY3Rpb249dmlld3ZhY2J5am9ibGlzdCZzZWFyY2hzb3J0PWdlb2Rpc3RhbmNlJmpvYmxpc3Rfdmlld192YWM9MTgyMzg5NiZvd25lcj01MDcwMDAwJm93bmVydHlwZT1mYWlyJnBhZ2VjbGFzcz1Kb2JzJnVzZXJzZWFyY2hjb250ZXh0PTE2NjM1MjYxJnJlcXNpZz0xNjY4NzkzMzEzLTQxYzM2MjVlYmMwY2Y3ZmJkMmIxMzdhMTM3NDVhYjgxOGUwMWVjZTY=

**Reply by meltemgureller** [reply]
Thank you very much! I will search details.

**Comment by RaminFirouzi** [comment]
Title: Edge-intelligent humanoid cognitive robots based on the Internet of robotic things
Type: Essay

Overview: This essay investigates how the Internet of things can mitigate some of the robots’ challenges and also how it can be managed through the AGI principles.
